Initial version

iilyak opened the pull request.

iilyak-ibm committed:
Initial version

iilyak commented on src/couch_epi.erl:
```diff
@@ -0,0 +1,142 @@
+unsubscribe(SubscriptionId) ->
+    couch_epi_server:unsubscribe(SubscriptionId).
+
+%% The success typing is (atom() | tuple(),_,_,[any()],_) -> [any()]
```
Remove this comment message

davisp commented on src/couch_epi_codegen.erl:
```diff
@@ -0,0 +1,72 @@
+    lists:reverse(ASTForms).
+
+format_term(Data) ->
+    lists:flatten(io_lib:format("~p", [Data])).
```
That should be ~w instead of ~p shouldn't it?

iilyak replied:
Right. I'm actually considering wrapping Data with `term_to_binary` to make sure it works even if things like `pid()` or `ref()` are passed by accident.

davisp commented on src/couch_epi_data_source.erl:
```diff
@@ -0,0 +1,173 @@
+reload(Server) ->
+    gen_server:call(Server, reload).
+
+wait(Server) ->
```
Does anything call the wait functions? I see the fancy delayed init but I'm not sure I understand the reasoning for it.

iilyak replied:
It is used from test suite

davisp replied:
Cool. Should we wrap the export in an -ifdef(TEST) macro?

iilyak replied:
I remember someone is against this practice in ASF. I can do that though if you like.

davisp replied:
Enh, I wouldn't worry about it.

davisp commented on src/couch_epi_data_source.erl:
```diff
@@ -0,0 +1,173 @@
+read({file, FilePath}) ->
+    case file:consult(FilePath) of
+        {ok, Data} ->
+            {ok, hash(FilePath), Data};
```
Surely that should be hash(Data), no?

iilyak replied:
Indeed. It is so silly. I'm a terrible person.

iilyak replied:
Hm. Actually hash(FilePath) is correct. Since the hash function would read the file.

iilyak replied:
I'll rename `hash` -> `hash_of_file`.

davisp commented on src/couch_epi_server.erl:
```diff
@@ -0,0 +1,138 @@
+    end.
+
+add(Subscriptions, App, Key, MFA) ->
+    Subscription = make_ref(),
```
Rather than make_ref, you should pass the From through and use a monitor so that you can remove the subscription if the subscriber.

davisp commented:
The only thing I wish I could think of a better solution for is the checking to see if a data source file has changed. It seems a bit much to scan from disk every 0.1 seconds. Maybe just reducing that to something like 5s or similar would be enough? Or maybe use timestamps on files instead? I'm open to any ideas on that one.

davisp commented:
Also, I should say that this is good stuff. Quite simple and easy to follow.

iilyak commented:
> The only thing I wish I could think of a better solution for is the checking to see if a data source file has changed. 

It is actually implemented [see here](https://github.com/iilyak/couch_epi/blob/first_version/src/couch_epi_data_source.erl#L60) and [here](https://github.com/iilyak/couch_epi/blob/first_version/src/couch_epi_data_source.erl#L85)

iilyak commented:
> Also, I should say that this is good stuff. Quite simple and easy to follow.

Thank you.

iilyak-ibm committed:
Add `ignore_providers` option

iilyak-ibm committed:
Use `~w` instead of `~p` so we don't truncate data

iilyak-ibm committed:
Rename `hash(FilePath)` into `hashof_file(FilePath)`

iilyak-ibm committed:
Fix timing issue in the test suite

iilyak-ibm committed:
Use monitor to unsubscribe when subscriber die

davisp commented:
My point about the check to see if its been updated was about how that check works. Specifically that the file is currently being read ten times a second. For instance I was thinking about this line:

https://github.com/iilyak/couch_epi/blob/first_version/src/couch_epi_data_source.erl#L129

And maybe if we changed that from reading the entire file to comparing the mtime. Granted that has all the standard issues of mtime being whacky in weird cases.

iilyak commented:
It will be read on every `tick` message which is [5 sec by default](https://github.com/iilyak/couch_epi/blob/first_version/src/couch_epi_data_source.erl#L16). Different value could be passed as `{interval, MSec}` option to start_link function. I can add support for mtime but I would rather have it configurable and disabled by default since mtime is usually disabled on SSD.

iilyak-ibm committed:
Rename couch_epi:all into couch_epi:dump

davisp commented:
I saw it was configurable, I just looked at the examples in the README that showed 100ms for the tick interval. 5s is probably fine.

davisp commented:
Also, good call on it being disabled. I'd call it good at the default of 5s for now and we can make a note about it in the docs

iilyak-ibm committed:
Handle stop message

iilyak-ibm committed:
Remove modules from dispatch on termination

iilyak-ibm committed:
Add all/any convinence funs to couch_epi

iilyak-ibm committed:
Fix handling of `stop` message

iilyak-ibm committed:
Add couch_epi_data

iilyak-ibm committed:
Document `ignore_providers` option

iilyak-ibm committed:
Add childspec helper to _data_source and _functions

iilyak-ibm committed:
Update documentation

iilyak-ibm committed:
Use code_change to trigger generation of a module

Previously we did periodic check of hashes for every provider module.
We switch to code_change use. Which would be called if specify modules
argument in child spec of the supervisor from where we register our
provider module.

We rely on this bit "Modules should be a list with one element [Module],
where Module is the name of the callback module, if the child process
is a supervisor, gen_server or gen_fsm. If the child process is a
gen_event, Modules should be dynamic." from
http://www.erlang.org/doc/design_principles/sup_princ.html

iilyak-ibm committed:
Rewrite test suite to test new functionality

iilyak-ibm committed:
Remove unused function

iilyak-ibm committed:
Fix compilation warnings
